source/extensions/filters/network/thrift_proxy/router/shadow_writer_impl.h

```diff
@@ -0,0 +1,262 @@
+    return FilterStatus::Continue;
+  }
+  FilterStatus transportEnd() override {
+    ASSERT(metadata_ != nullptr);
+    complete_ = true;
+    return FilterStatus::Continue;
+  }
+
+  // DecoderCallbacks
+  DecoderEventHandler& newDecoderEventHandler() override { return *this; }
+  bool passthroughEnabled() const override { return false; }
+
+  DecoderPtr decoder_;
+  Buffer::OwnedImpl response_buffer_;
+  Buffer::OwnedImpl upstream_buffer_;
+  MessageMetadataSharedPtr metadata_;
+  absl::optional<bool> success_;
+  bool complete_ : 1;
+  bool first_reply_field_ : 1;
+};
+using NullResponseDecoderPtr = std::unique_ptr<NullResponseDecoder>;
+
+// Adapter from NullResponseDecoder to UpstreamResponseCallbacks.
+class ShadowUpstreamResponseCallbacksImpl : public UpstreamResponseCallbacks {
+public:
+  ShadowUpstreamResponseCallbacksImpl(NullResponseDecoder& response_decoder)
+      : response_decoder_(response_decoder) {}
+
+  void startUpstreamResponse(Transport&, Protocol&) override {}
+  ThriftFilters::ResponseStatus upstreamData(Buffer::Instance& buffer) override {
+    return response_decoder_.upstreamData(buffer);
+  }
+  MessageMetadataSharedPtr responseMetadata() override {
+    return response_decoder_.responseMetadata();
+  }
+  bool responseSuccess() override { return response_decoder_.responseSuccess(); }
+
+private:
+  NullResponseDecoder& response_decoder_;
+};
+using ShadowUpstreamResponseCallbacksImplPtr = std::unique_ptr<ShadowUpstreamResponseCallbacksImpl>;
+
+class ShadowWriterImpl;
+
+class ShadowRouterImpl : public ShadowRouterHandle,
+                         public RequestOwner,
+                         public Tcp::ConnectionPool::UpstreamCallbacks,
+                         public Upstream::LoadBalancerContextBase,
+                         public Event::DeferredDeletable,
+                         public LinkedObject<ShadowRouterImpl> {
+public:
+  ShadowRouterImpl(ShadowWriterImpl& parent, const std::string& cluster_name,
+                   MessageMetadataSharedPtr& metadata, TransportType transport_type,
+                   ProtocolType protocol_type);
+  ~ShadowRouterImpl() override = default;
+
+  bool createUpstreamRequest();
+  void maybeCleanup();
+  void resetStream() {
+    if (upstream_request_ != nullptr) {
+      upstream_request_->releaseConnection(true);
+    }
+  }
+
+  // ShadowRouterHandle
+  void onRouterDestroy() override;
+  bool waitingForConnection() const override;
+  RequestOwner& requestOwner() override { return *this; }
+
+  // RequestOwner
+  Tcp::ConnectionPool::UpstreamCallbacks& upstreamCallbacks() override { return *this; }
+  Buffer::OwnedImpl& buffer() override { return upstream_request_buffer_; }
+  Event::Dispatcher& dispatcher() override;
+  void addSize(uint64_t size) override { request_size_ += size; }
+  void continueDecoding() override {
+    if (pending_callbacks_.empty()) {
+      return;
+    }
+
+    for (auto& cb : pending_callbacks_) {
+      cb();
+    }
```

Comment:
Should we clear pending_callbacks_?

Reply:
If you've reached this point, nothing else will be pushed into `pending_callbacks_`.  Thus my thinking was let it be cleaned up when `~ShadowRouterImpl()` happens.